### PR TITLE
feat(api): check ms trophy urls exist

### PR DESCRIPTION
**Pull request by ojeytonwilliams**
- feat(client): warn if server cannot find trophy
- feat: add function to obtain ms api url
- feat(api): check trophy url on submission
- feat: upgrade warning to error
- refactor: combine error messages in single file

Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

The api checks the Microsoft Learn link and warns the user if their submission could not be confirmed to exist.

This is performed by the server, not the client, because our domain is not an allowed origin.

After https://github.com/freeCodeCamp/freeCodeCamp/pull/51008 the remaining task is

- Validate the form of the link on the server

### CODE REVIEW

File: api-server/src/server/boot/challenge.js [hunk 1]
```diff
@@ -372,6 +374,18 @@ async function projectCompleted(req, res, next) {
     }
   }
 
+  const isMSTrophyProject = completedChallenge.challengeType === 18;
+  let isTrophyMissing = false;
+  if (isMSTrophyProject) {
+    try {
+      const mSLearnAPIUrl = getApiUrlFromTrophy(completedChallenge.solution);
+      isTrophyMissing = mSLearnAPIUrl ? !(await fetch(mSLearnAPIUrl)).ok : true;
```
[the remark this answers is not in the document]
Comment (ojeytonwilliams) [reply api-server/src/server/boot/challenge.js:382]: Good call, but it turns out that endpoint doesn't exist. I don't know why, but it only responds to `GET` requests.

File: api-server/src/server/boot/challenge.js [hunk 2]
```diff
@@ -372,6 +374,18 @@ async function projectCompleted(req, res, next) {
...
+    } catch {
+      isTrophyMissing = true;
+      log('Error verifying trophy');
```
[the remark this answers is not in the document]
Comment (ojeytonwilliams) [reply api-server/src/server/boot/challenge.js:385]: Works for me

**Review by naomi-lgbt — APPROVED** [review]
Logic looks good to me.

File: api-server/src/server/boot/challenge.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (ojeytonwilliams) [reply api-server/src/server/boot/challenge.js:382]: Yeah, it's weird. My guess is that their server doesn't set them automagically (unlike express and fastify).

**Review by moT01 — COMMENTED** [review]
It looks this is still saving invalid trophy URL's to the database - e.g. submitting `{"id":"647f867a07d29547b3bee1bc","challengeType":18,"solution":"https://freecodecamp.org"}`. So campers will still be getting credit for completing the challenge. Not sure how to best handle that. Early return if isTrophyMissing? or maybe skip the user updates if it's missing.

**Comment by ojeytonwilliams** [comment]
@moT01 my thinking was that if this detects an invalid url it's possible that's a false positive (at least in principle). That's why it just warns the learner if their submission might be wrong, so they can fix it.

That said, if it's better to be strict, I'll change it so that invalid URLs are not stored in the DB.

**Comment by moT01** [comment]
I like using their API to see if they've earned the trophy - but I'm not sure how much we can rely on that. So I don't have that much of an issue saving it to completedChallenges if that gives a not okay response or fails for whatever reason. But I think we need to make sure that the URL they are submitting matches the pattern we expect - so they don't get credit for submitting random URL's.

**Comment by ojeytonwilliams** [comment]
Sounds good. I'll add apply the same validation as we do on the client.

**Review by moT01 — APPROVED** [review]
LGTM 🎉 Thanks again for working on this @ojeytonwilliams 👍

**Pull request merged**